h2/src/main/org/h2/engine/Session.java

```diff
@@ -1768,6 +1748,64 @@ public void markTableForAnalyze(Table table) {
         tablesToAnalyze.add(table);
     }
 
+    @Override
+    public void onRollback(MVMap<Object, VersionedValue> map, Object key,
+                           VersionedValue existingValue,
+                           VersionedValue restoredValue) {
+        // Here we are relying on the fact that map which backs table's primary index
+        // has the same name as the table itself
+        MVTableEngine.Store store = database.getMvStore();
+        if(store != null) {
+            MVTable table = store.getTable(map.getName());
+            if (table != null) {
+                long recKey = ((ValueLong)key).getLong();
+                Row oldRow = getRowFromVersionedValue(table, recKey, existingValue);
+                Row newRow = getRowFromVersionedValue(table, recKey, restoredValue);
+                table.fireAfterRow(this, oldRow, newRow, true);
+
+                if (table.getContainsLargeObject()) {
+                    if (oldRow != null) {
+                        for (int i = 0, len = oldRow.getColumnCount(); i < len; i++) {
+                            Value v = oldRow.getValue(i);
+                            if (v.isLinkedToTable()) {
+                                removeAtCommit(v);
+                            }
+                        }
+                    }
+                    if (newRow != null) {
+                        for (int i = 0, len = newRow.getColumnCount(); i < len; i++) {
+                            Value v = newRow.getValue(i);
+                            if (v.isLinkedToTable()) {
+                                removeAtCommitStop(v);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private static Row getRowFromVersionedValue(MVTable table, long recKey,
+                                                VersionedValue versionedValue) {
+        Object value = versionedValue == null ? null : versionedValue.value;
+        Row result = null;
+        if (value != null) {
+            Row result11;
```

Comment:
Looks like `result` and `result11` can be merged into one variable.

However, it's may be even better just to return `null` immediately if `value == null` without variable assignment.

Reply:
You got me here... I should’ve read result of automated method inlining :sad: